cdap-common/src/main/java/co/cask/cdap/common/zookeeper/ZKExtOperations.java

```diff
@@ -353,58 +274,293 @@ public void onFailure(Throwable t) {
+   * @see #createOrSet(ZKClient, String, Supplier, Codec, int, List)
+   * @see #setOrCreate(ZKClient, String, Supplier, Codec, int)
+   */
+  private static <T> ListenableFuture<T> createOrSetWithRetry(final boolean createFirst, final ZKClient zkClient,
+                                                              final String path, final Supplier<T> dataSupplier,
+                                                              final Codec<T> codec, @Nullable final Iterable<ACL> acls,
+                                                              final int maxRetry) {
+    final SettableFuture<T> resultFuture = SettableFuture.create();
+    final AtomicInteger failures = new AtomicInteger(0);
+
+    Futures.addCallback(
+      doCreateOrSet(createFirst, zkClient, path, dataSupplier, codec, acls),
+      new FutureCallback<T>() {
+        @Override
+        public void onSuccess(T result) {
+          resultFuture.set(result);
+        }
+
+        @Override
+        public void onFailure(Throwable t) {
+          if (failures.getAndIncrement() < maxRetry) {
+            Futures.addCallback(doCreateOrSet(createFirst, zkClient, path, dataSupplier, codec, acls),
+                                this, Threads.SAME_THREAD_EXECUTOR);
+          } else {
+            resultFuture.setException(t);
+          }
+        }
+      }, Threads.SAME_THREAD_EXECUTOR
+    );
+
+    return resultFuture;
+  }
+
+  /**
+   * Tries create or set the content of a node. If {@code createFirst} is {@code true}, it will first try to create;
+   * and if it failed with NodeExists error, then will try to perform set.
+   * If {@code createFirst} is {@code false}, then it will first try to set and if it failed with NoNode error,
+   * then will try to perform create.
+   *
+   * @param createFirst if {@code true}, try create and then set; if {@code false}, try set and then create.
+   * @param zkClient The ZKClient to perform the operations.
+   * @param path The path in ZK.
+   * @param dataSupplier The supplier to provide the content to be set to the node.
+   * @param codec A {@link Codec} for serializing the data into byte array.
+   * @param acls The access control list to set on the node.
+   * @param <T> type of the data to set to the node
+   * @return A {@link ListenableFuture} that will be completed when node is created or data is set.
+   *         The future will carry the actual content being set into the node.
+   */
+  private static <T> ListenableFuture<T> doCreateOrSet(final boolean createFirst, final ZKClient zkClient,
+                                                       final String path, final Supplier<T> dataSupplier,
+                                                       final Codec<T> codec, @Nullable final Iterable<ACL> acls) {
+    final SettableFuture<T> resultFuture = SettableFuture.create();
+    final Supplier<ListenableFuture<T>> futureSupplier = createFutureSupplier(dataSupplier);
```

Comment:
The private `create` method that we are going to call takes a `Supplier<ListenableFuture<T>>`, so we need to adapt the `Supplier<T>` we got in this method to that.

As for why the private `create` method takes a `Supplier<ListenableFuture<T>>` is because that method is also used to support the `createOrGetAndSet` case.